source/use-cases/sensitive-data-encryption.txt

```diff
@@ -0,0 +1,312 @@
+- Nullam imperdiet lorem vitae vulputate lacinia.
+- Donec eget velit tincidunt, gravida diam ac, efficitur lacus.
+
+A. Create a Local Master Key
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Aenean eu consequat lorem. Ut posuere est sed sodales pharetra. Cras
+volutpat, massa laoreet varius dictum, leo odio porttitor ante, nec
+auctor tortor orci et mi. Maecenas tempor, lacus vehicula molestie
+pulvinar, ante eros faucibus odio, sed consequat quam tellus vel arcu.
+Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere
+cubilia Curae; Nunc non interdum purus, ultricies laoreet tortor.
+
+B. Define a JSON Schema
+~~~~~~~~~~~~~~~~~~~~~~~
+
+MongoDB uses JSON Schema to to validate data which is written to a collection.
+Within the context of Field Level Encryption, we use JSON Schema to specify several
+configuration details. For instance, we can specify the form of encryption we want to use 
+- deterministic or random. We can also specify the BSON Type of the field we are encrypting. 
+In addition, we can specify the key we want to use. This is particularly important, because
+it means that we can use different keys for different fields.
+
+In the following example, we will go over creation of a schema for a collection which represents patient data
+in a medical office. Some fields use deterministic encryption, and others use random (probabilistic encryption). 
+
+The structure for our schema will be as follows:
+
+.. code::
+
+    fullName: String
+    address: Encrypted String using random encryption
+    ssn: Encrypted Int using deterministic encryption
+    last4SSN: Int
+    patientInfo: Object
+    patientInfo.phone Encrypted Int using deterministic encryption
+    patientInfo.provider String
+    medicalRecords: Encrypted array of documents using random encryption
+
+Explanation of encryption choices:
+``````````````````````````````````
+* address is randomly encrypted because it is more secure for strings to be randomly encrypted to prevent hackers from decrypting by building a cipher-dictionary
+* ssn is deterministically encrypted because deterministic encryption provides filtering ability 
+* patientInfo.phone is deterministically encrypted for the same reason as ssn, and also it provides an example of encryption with a nested document
+* medicalRecords are randomly encrypted because entire array's are only encryptable with random encryption
+
+
+.. tabs::
+
+   tabs:
+
+     - id: java-jsonschema-generation
+       name: "Java"
+       content: |
+
+         The following code creates a collection using the schema above and inserts it into a database using MongoDB.
+         For the "Define a JSON Schema" section we've abstracted the createSecureClient, to focus on creation of a schema.
+         But that function uses getJSONSchema and passes it a dataKey id in order to create a client which uses the JSONSchema
+         we want to use. 
+
+         .. code-block:: java
+
+            import com.mongodb.*;
```

Comment:
Quick question: is this import necessary?

Reply:
It throws the following error without it:
`le/fle/CleanFleApp.java:120: error: cannot find symbol
                var keyVaultSettings = ClientEncryptionSettings.builder()
                                       ^
  symbol:   variable ClientEncryptionSettings
  location: class CleanFleApp
/Users/mohammad.hunan/Code/docs-fle/examples/javaexample/src/main/java/com/example/fle/CleanFleApp.java:121: error: cannot find symbol
                                .keyVaultMongoClientSettings(MongoClientSettings.builder()
                                                             ^
  symbol:   variable MongoClientSettings
  location: class CleanFleApp
/Users/mohammad.hunan/Code/docs-fle/examples/javaexample/src/main/java/com/example/fle/CleanFleApp.java:132: error: cannot find symbol
                                AutoEncryptionSettings.builder()
                                ^
  symbol:   variable AutoEncryptionSettings
  location: class CleanFleApp
/Users/mohammad.hunan/Code/docs-fle/examples/javaexample/src/main/java/com/example/fle/CleanFleApp.java:141: error: cannot find symbol
                var clientSettings = MongoClientSettings.builder()
                                     ^
  symbol:   variable MongoClientSettings
  location: class CleanFleApp
Note: /Users/mohammad.hunan/Code/docs-fle/examples/javaexample/src/main/java/com/example/fle/CleanFleApp.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
4 errors
`

Reply:
Ok makes sense, it's in a part of the app that isn't shown.